### ISSUE TITLE
Improved and more secure API Key handling

### ISSUE THREAD

**Issue by rkg-mm** [DependencyTrack/dependency-track#2552]
### Current Behavior

Each team gets an API key automatically. All API keys are visible for admins in the web interface (therefore must be stored somewhere in plaintext or only encrypted to be revertable).

### Proposed Behavior

1. By default a Team should not have API keys
2. It shall be able to generate API keys by admins, ideally with an additional comment or name, to document the purpose
3. API keys should not be visible in plain text except once after generation
4. API keys should be stored as a secret with proper one-way hashing (basically it's a secret and must be handled like one!)
5. (Ideally) Creation Date and Last usage date should be shown next to the API key for the admin to be able to clear up

### Checklist

- [X] I have read and understand the [contributing guidelines](https://github.com/DependencyTrack/dependency-track/blob/master/CONTRIBUTING.md#filing-issues)
- [X] I have checked the [existing issues](https://github.com/DependencyTrack/dependency-track/issues) for whether this enhancement was already requested

**Comment by valentijnscholten** [comment]
I agree these improvements are welcome/needed, but probably should be moved to https://github.com/stevespringett/Alpine ?

**Comment by rkg-mm** [comment]
If this is part of the alpine framework, then yes :D

**Comment by mprencipe** [comment]
Alpine doesn't automatically create API keys for teams, it's parametrized:
https://github.com/stevespringett/Alpine/blob/master/alpine-infra/src/main/java/alpine/persistence/AlpineQueryManager.java#L559

Dependency-Track invokes the inherited AlpineQueryManager#createTeam method with the "true" parameter which generates the API key: https://github.com/DependencyTrack/dependency-track/blob/master/src/main/java/org/dependencytrack/resources/v1/TeamResource.java#L135

The simplest fix would be changing true to false.

**Comment by nscuro** [comment]
@mprencipe That sounds like a sensible thing to do. Do you fancy raising a PR for this?

@rkg-mm:

> 2. It shall be able to generate API keys by admins, ideally with an additional comment or name, to document the purpose

Comments to document the purpose is coming in v4.11, as per https://github.com/DependencyTrack/frontend/pull/768.

> 6. (Ideally) Creation Date and Last usage date should be shown next to the API key for the admin to be able to clear up

Timestamps to track creation and "last used" timestamps are coming in v4.11, as per https://github.com/DependencyTrack/frontend/pull/768.

> 7. Also consider https://github.com/DependencyTrack/dependency-track/issues/2543

This was shipped in v4.9.

**Comment by mprencipe** [comment]
Sure, I can raise a PR.

**Comment by nscuro** [comment]
This has been implemented fully, with the majority of it having shipped in 4.13.0: https://docs.dependencytrack.org/changelog/#v4-13-0

Thanks @msymons for identifying this as closeable.
